File: Linked List /C++/7. 328. Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == NULL || head ->next == NULL)
        {
            return head;
        }

        ListNode* odd;
        ListNode* even;
        ListNode* odd_head = odd = head;
        ListNode* even_head = even = head->next;

        while (even != NULL && even->next != NULL)
        {
            odd->next = even->next;
            odd = odd->next;
            even->next = odd->next;
            even = even->next;
        }
        odd->next = even_head;
        return odd_head;
    }
};

File: Linked List /Python/7. 328. Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        # if list is empty or having a single node:
        if not head or not head.next:
            return head
        
        # separating the list into odd list and even list and combine them
        odd_head = odd = head
        even_head = even = head.next

        while even and even.next:
            odd.next = even.next
            odd = odd.next
            even.next = odd.next
            even = even.next

        # combine the odd list with even list and return it
        odd.next = even_head
        return odd_head
